src/flownet/ahm/_run_ahm.py

```diff
@@ -215,6 +216,99 @@ def _get_distribution(
     return df
 
 
+def _constrain_using_well_logs(
+    porv_poro_trans_dist_values: pd.DataFrame,
+    data: np.ndarray,
+    network: NetworkModel,
+    measurement_type: str,
+    config: ConfigSuite.snapshot,
+) -> pd.DataFrame:
+    """
+    Function to constrain permeability and porosity distributions of flow tubes by using 3D kriging of
+    porosity and permeability values from well logs.
+
+    Args:
+        porv_poro_trans_dist_values: pre-constraining dataframe
+        data: well log data
+        network: FlowNet network model
+        measurement_type: 'prorosity' or 'permeability' (always log10)
+        config: FlowNet configparser snapshot
+
+    Returns:
+        Well-log constrained "porv_poro_trans_dist_values" DataFrame
+
+    """
+    n = config.flownet.constraining.kriging.n
+    n_lags = config.flownet.constraining.kriging.n_lags
+    anisotropy_scaling_z = config.flownet.constraining.kriging.anisotropy_scaling_z
+    variogram_model = config.flownet.constraining.kriging.variogram_model
+
+    if measurement_type == "permeability":
+        data[:, 3] = np.log10(data[:, 3])
+
+    variogram_parameters: Optional[Dict] = None
+    if measurement_type == "permeability":
+        variogram_parameters = dict(
+            config.flownet.constraining.kriging.permeability_variogram_parameters
+        )
+    elif measurement_type == "porosity":
+        variogram_parameters = dict(
+            config.flownet.constraining.kriging.porosity_variogram_parameters
+        )
+
+    if not variogram_parameters:
+        variogram_parameters = None
+
+    k3d3_interpolator, ss3d_interpolator = kriging.execute(
+        data,
+        n=n,
+        n_lags=n_lags,
+        variogram_model=variogram_model,
+        variogram_parameters=variogram_parameters,
+        anisotropy_scaling_z=anisotropy_scaling_z,
+    )
+
+    parameter_min_kriging = k3d3_interpolator(
+        network.connection_midpoints
+    ) - ss3d_interpolator(network.connection_midpoints)
+
+    parameter_max_kriging = k3d3_interpolator(
+        network.connection_midpoints
+    ) + ss3d_interpolator(network.connection_midpoints)
```

Comment:
`ss3d_interpolator` gives the variance (sigma squared). should you not use the standard deviation to define the min/max (ref https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule)?

Reply:
Good catch. Let's go for

<img src="https://latex.codecogs.com/svg.latex?\pm%20\sqrt{\sigma^{2}}" />,

which is the 95% confidence interval. Agreed?

Reply:
I guess we can start there, at least. For porosity it should give a much broader range (since the variance is << 1).

Reply:
It might also explain why it worked _not so well_... 😄 😃